sql/core/src/main/scala/org/apache/spark/sql/execution/stat/StatFunctions.scala

```diff
@@ -77,4 +78,42 @@ private[sql] object StatFunctions {
       })
     counts.cov
   }
+
+  /** Generate a table of frequencies for the elements of two columns. */
+  private[sql] def crossTabulate(df: DataFrame, col1: String, col2: String): DataFrame = {
+    val tableName = s"${col1}_$col2"
+    val distinctCol2 = df.select(col2).distinct.collect().sortBy(_.get(0).toString)
```

Comment:
I'm happy to implement it both ways. Check my first commit. If you both think that's ok, or a combination of both ideas is better, I'd be happy to implement it. I think my first implementation had two passes as well. I think you might need two passes two pivot properly. Maybe we can do something smarter. cc @mengxr